src/Algebra/Graph/Relation/Symmetric.hs

```diff
@@ -11,27 +11,497 @@
+--
+-- @
+-- biclique []      []      == 'empty'
+-- biclique [x]     []      == 'vertex' x
+-- biclique []      [y]     == 'vertex' y
+-- biclique [x1,x2] [y1,y2] == 'edges' [(x1,y1), (x1,y2), (y1,x1), (y1,x2), (x2,y1), (x2,y2), (y2,x1), (y2,x2)]
+-- biclique xs      ys      == 'connect' ('vertices' xs) ('vertices' ys)
+-- @
+biclique :: Ord a => [a] -> [a] -> SymmetricRelation a
+biclique xs = SymmetricRelation . R.symmetricClosure . R.biclique xs
+
+-- TODO: Optimise.
+-- | The /star/ formed by a centre vertex connected to a list of leaves.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- star x []    == 'vertex' x
+-- star x [y]   == 'edges' [(x,y), (y,x)]
+-- star x [y,z] == 'edges' [(x,y), (x,z), (y,x), (y,z)]
+-- star x ys    == 'connect' ('vertex' x) ('vertices' ys)
+-- @
+star :: Ord a => a -> [a] -> SymmetricRelation a
+star x [] = vertex x
+star x ys = connect (vertex x) (vertices ys)
+
+-- | The /stars/ formed by overlaying a list of 'star's. An inverse of
+-- 'adjacencyList'.
+-- Complexity: /O(L * log(n))/ time, memory and size, where /L/ is the total
+-- size of the input.
+--
+-- @
+-- stars []                      == 'empty'
+-- stars [(x, [])]               == 'vertex' x
+-- stars [(x, [y])]              == 'edges' [(x,y), (y,x)]
+-- stars [(x, ys)]               == 'star' x ys
+-- stars                         == 'overlays' . 'map' ('uncurry' 'star')
+-- stars . 'adjacencyList'         == id
+-- 'overlay' (stars xs) (stars ys) == stars (xs ++ ys)
+-- @
+stars :: Ord a => [(a, [a])] -> SymmetricRelation a
+stars as = SymmetricRelation $ RI.Relation (Set.fromList vs) (Set.fromList es)
+  where
+    vs = concatMap (uncurry (:)) as
+    es = [ (x, y) | (x, ys) <- as, y <- ys ] ++ [ (y, x) | (x, ys) <- as, y <- ys ]
+
+-- | The /tree graph/ constructed from a given 'Tree.Tree' data structure.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- tree (Node x [])                                         == 'vertex' x
+-- tree (Node x [Node y [Node z []]])                       == 'path' [x,y,z]
+-- tree (Node x [Node y [], Node z []])                     == 'star' x [y,z]
+-- tree (Node 1 [Node 2 [], Node 3 [Node 4 [], Node 5 []]]) == 'edges' [(1,2), (1,3), (2,1), (3,1), (3,4), (3,5), (4,3), (5,3)]
+-- @
+tree :: Ord a => Tree.Tree a -> SymmetricRelation a
+tree (Node x []) = vertex x
+tree (Node x f ) = star x (map rootLabel f)
+    `overlay` forest (filter (not . null . subForest) f)
+
+-- | The /forest graph/ constructed from a given 'Tree.Forest' data structure.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- forest []                                                  == 'empty'
+-- forest [x]                                                 == 'tree' x
+-- forest [Node 1 [Node 2 [], Node 3 []], Node 4 [Node 5 []]] == 'edges' [(1,2), (1,3), (2,1), (3,1), (4,5), (5,4)]
```

Comment:
No need to duplicate the edges in this example.

Reply:
Already answered